nntrainer/tensor/tensor_dim.h

```diff
@@ -323,13 +360,36 @@ class TensorDim {
    */
   void reverse();
 
+  /**
+   * @brief Get the Effective Dimension of the current
+   * @note dynamic dimension is returned as -1
+   *
+   * @param dynamic if dimension has to be considering dynamic set this to ture
+   * @return std::vector<unsigned int> integer vector
+   */
+  std::vector<int> getEffectiveDimension(bool dynamic = false) const;
+
+  /**
+   * @brief check if tensor is dynamic
+   *
+   * @return true any of dyn_dim_flag is set
+   * @return false none of dyn_dim_flag is set
+   */
+  bool is_dynamic() const { return dyn_dim_flag.any(); }
+
 private:
   /**
    * @brief reset length
    *
    */
   void resetLen();
 
+  std::bitset<MAXDIM> eff_dim_flag; /**< dimension bit flag to define effective
+          dimension size */
+
+  std::bitset<MAXDIM> dyn_dim_flag; /**< dimension bit flag to define
```

Comment:
Can dimensions other than batch be dynamic? 
Maybe all dimensions can be dynamic but are there are any other possibilities?

Reply:
Yes, all dimension can be dynamic.

Reply:
If the dynamic dimensions are other than the batch dimension, we must add a way to ensure that the calculation of output must work independent of it, else fail. Please add a TODO related to this.

Reply:
food for thought - set `-1` to the values which are dynamic directly in the dimension. Just like a value of 0 inside dimension means it's not properly set yet, `-1` would mean it's not yet decided (this dimension's actual value would be 1 or more). Using this dimension for calculation purposes will work, but this dimension can't be used to create a Tensor or allocate any memory related to it.

Reply:
Yeah, I am thinking that the reason they have a dynamic dimension is similar. Having both (effective, dynamic) dimension has advantage that when dimension is not set, there is a explicit default dimension. -1 could be used for our purpose as well.

> If the dynamic dimensions are other than the batch dimension, we must add a way to ensure that the calculation of output must work independent of it, else fail. Please add a TODO related to this.

Added a todo regarding your concern.